harness/tests/experiment/pytorch/test_pytorch_context.py

```diff
@@ -0,0 +1,52 @@
+import unittest
+
+import torch
+
+from determined import pytorch
+from determined.common.check import CheckFailedError
+from determined.errors import InternalException
+from tests.experiment.fixtures import pytorch_onevar_model
+
+
+class TestPyTorchContext(unittest.TestCase):
+    def setUp(self) -> None:
+        self.config = {"hyperparameters": {"global_batch_size": 4, "dataloader_type": "determined"}}
+        self.context: pytorch.PyTorchTrialContext = pytorch.PyTorchTrialContext.from_config(
+            self.config
+        )
+
+    def test_from_config(self) -> None:
+        assert isinstance(self.context, pytorch.PyTorchTrialContext)
+
+        trial = pytorch_onevar_model.OneVarTrial(self.context)
+
+        train_ds = trial.build_training_data_loader()
+        for epoch_idx in range(3):
+            for batch_idx, batch in enumerate(train_ds):
+                metrics = trial.train_batch(batch, epoch_idx, batch_idx)
+                # Verify the training is correct.
+                pytorch_onevar_model.OneVarTrial.check_batch_metrics(metrics, batch_idx)
+
+        eval_ds = trial.build_validation_data_loader()
+        for batch in eval_ds:
+            metrics = trial.evaluate_batch(batch)
+
+    def test_average_gradients(self) -> None:
+        self.assertRaises(CheckFailedError, self.context._average_gradients, None, 0)
+        self.assertIsNone(self.context._average_gradients(None, 1))
+
+    def test_training_not_started(self) -> None:
+        self.assertRaises(InternalException, self.context.is_epoch_start)
+        self.assertRaises(InternalException, self.context.is_epoch_end)
+        self.assertRaises(InternalException, self.context.current_train_batch)
+        self.assertRaises(InternalException, self.context.current_train_epoch)
```

Comment:
Whoa, we should not be raising an InternalException here.  I never realized this was happening.  InternalException should be reserved for logic bugs, not "user used a public method before we wanted them to".

Testing for these exceptions is fine, but can you change them to be normal python `RuntimeErrors` instead of det.errors.InternalExceptions?

Reply:
It would have to be `Exception`, because that's what `errors.InternalException` inherits from.

Reply:
I also don't think using `errors.InternalException` is wrong: after all, it is our code, and we should be as specific as possible.

Reply:
Why would it have to be an Exception?

Returning an `InternalException` as a result of a user using an API when they are not allowed to is definitely wrong.  InternalException is like an HTTP 500: it means you've done nothing wrong, but you've uncovered a bug in somebody else's code.

In these cases, the user is just not allowed to call those methods during `Trial.__init__` and it indicates no internal error if the user tries to call them at the wrong time.

Reply:
I'm not arguing about `InternalException` being the wrong thing for these methods to raise. I'm saying that that's the exception that these methods raise, and `InternalException` inherits from `Exception`.
Using `self.assertRaises(RuntimeError, blah)` does not work.

Reply:
actually, should we _ever_ throw `InternalException`? in all current uses, it's more like an invalid input/usage error. philosophically, does it even make sense? when would we catch known Runtime errors and throw up a 500? 

maybe we should just change that exception to something else, like InvalidUsageException or something. @rb-determined-ai

Reply:
> I'm saying that that's the exception that these methods raise

And I am saying that it's clearly wrong for a public API to raise an "InternalError".  I'm asking you to change the functions, not just the test.

> maybe we should just change that exception to something else, like InvalidUsageException or something.

I'm fine with this too.  Just public methods should never raise an InternalError when used wrong by the user.  InternalError would mean that we had a bug, while InvalidUsageException or RuntimeError would indicate that the user has a bug (arguably, the former is clearer about that than the latter).

Reply:
> And I am saying that it's clearly wrong for a public API to raise an "InternalError". I'm asking you to change the functions, not just the test.
I agree with the first sentence, and I'd be happy to make the change you're asking for, but that change is outside the scope of this PR and associated tickets.

Reply:
let's just change the exception in the code and tests to InvalidUsageException. we can do it in a separate PR, and leave this as is for now, since without changing the code, this test is technically accurate. agree? @rb-determined-ai @drh-determined-ai

Reply:
I agree with @azhou-determined and will gladly be the assignee for the task.